### ISSUE TITLE
Investigate support of Tek4010 graphics mode/terminal

### ISSUE THREAD

**Issue by WSLUser** [dankamongmen/notcurses#2068]
Randomly discovered an OSS in use terminal that performs emulation of Tektronix 4010, Tektronix 4013, Tektronix 4014, Tektronix 4015 and ARDS and offers graphic mode support that not even Xterm supports.  https://github.com/rricharz/Tek4010

It does appear mostly geared to users who prefer a more retro look so it may be akin to the support of the Linux console, albeit with slightly better graphics available.  Main concern is really with 

> Standard resolution of window 1024 x 780 points

and of course I'd want that scaled up to HD resolution at minimum. Also my discovery came from reviewing features Mintty added within the past year (I haven't really kept up on all the stuff that was added) and they apparently support the graphics mode, so if notcurses does support this, it needs to identify support via query, not terminfo.

**Comment by dankamongmen** [comment]
so far as i'm aware, we ought support Tektronix fairly well; terminfo has a good deal of information about them, so they're pretty clearly well-supported therein.

a few things to check:

>                                hue_lightness_saturation    hls       hl     terminal uses only HLS color notation (Tektronix)

so long as they don't advertise RGB, this is no problem, except for setting the palette, which we would presumably need do with HLS semantics.

> If the terminal overstrikes (rather than clearing a position when a character is struck over) then it should have the os capability.  If the terminal is a printing terminal, with no soft copy unit, give it both hc and os (os  applies  to  storage scope terminals, such as TEKTRONIX 4010 series, as well as hard copy and APL terminals.)

i have no intention of supporting hard copy terminals, so maybe we just check for `os` and exit if it's present, much as we do if `am` is missing (i think).

> If there  are  parameterized  local  motions (e.g., move n spaces to the right) these can be given as cud, cub, cuf, and cuu with a single parameter indicating how many spaces to move.  These are primarily useful if the  terminal  does  not  have cup, such as the TEKTRONIX 4025.

this is a problem, as we exit immediately without `cup`. i could of course synthesize it from `vpa`+`hpa` or, if those are also missing, `cu[dbfu]`, but i don't have code to do so at this time.

> This  is  also used for the TEKTRONIX 4025, where smcup sets the command character to be the one used by terminfo.  If the smcup sequence will not restore the screen after an rmcup sequence is output (to the state prior to outputting rmcup), specify nrrmc.

it sounds to me like this is saying that on the 4025, you always need emit `smcup`, at least to use `cup`? if that's the case, we would want to add code that disallows `NCOPTION_NO_ALTERNATE_SCREEN` when `nrrmc` is defined, or at least ignores it.

> Tektronix-like  terminals have a predefined set of N colors (where N is usually 8), and can set character-cell foreground and background characters independently, mixing them into N * N color-pairs.

good; this is what we already assume. HP terminals use colorpairs, which we do not yet support (but might in the future).

> If the terminal has a settable command character, such as the TEKTRONIX 4025, this can be indicated with cmdch.

i have no idea what this means.

that exhausts the notes regarding TEKTRONIX in `terminfo(5)`. i can only assume that the changes listed above would serve to accommodate such terminals. i have no means to test them, so i can't say that with any certainty.

i think this is probably worth doing, but it's some pretty low-priority shit.

**Comment by j4james** [comment]
Note that Tektronix mode isn't like Sixel or ReGIS, which are designed to be mixed in with VT sequences. It's an entirely separate emulation. On the original DEC terminals, the screen would be cleared on entry and exit from Tektronix mode, and at least some of the modern implementations (e.g. XTerm and RLogin) display it in a completely separate window. I'm not sure there is a way you could integrate it meaningfully in notcurses, unless your intention is to treat it as a standalone concept.

**Comment by dankamongmen** [comment]
wait, i thought Tektronix was a manufacturer of terminals, not like a graphics protocol? i thought the other graphics protocol was regis vectors?

i'm not sure what to make of "DEC trerminals...Tektronix mode". i thought textronix was just a maker of terminals, competing with DEC?

**Comment by j4james** [comment]
Yeah, but the DEC graphics terminals were able to emulate Tektronix graphics (using DEC private mode 38), and modern terminal emulators that emulate DEC terminals will emulate their emulation of Tektronix. 😄

I just assumed notcurses was intended exclusively for VT derived terminals, so I thought it was the DEC Tektronix emulation you had in mind here, but if the idea is to support Tektronix terminals as a separate device type that makes a lot more sense. In which case you can disregard my ignorant ramblings.

**Comment by dankamongmen** [comment]
ok so it sounds like "textronix" is *both* a graphics protocol and a distinct terminal type, then? got it. i don't see myself implementing what sounds like a long-obsolete graphics protocol, but yeah, i'd like to work on these terminals otherwise. which ought just be a matter of making the changes outlined above.

**Comment by dankamongmen** [comment]
work list: 
* [ ] ~exit if `os` or `hc` capabilities are set~ *actually, we never emit backspaces, so this is immaterial*
* [ ] synthesize movement when `cup` is missing
* [ ] handle `nrrmc`

**Comment by WSLUser** [comment]
Friendly reminder now that 3.0 is out. Adding more terminal support would definitely be a great holiday present for people.
